### DISCUSSION TITLE
Which problem representation formats would you recommend as (i) well-specified and (ii) widely used?

### DISCUSSION THREAD

**Discussion by dlaehnemann** [coin-or/CoinUtils#149] [Q&A] [answered]
The Rust wrapper for ILP solvers [`rust_lp_modeler`](https://github.com/jcavat/rust-lp-modeler) supports using (COIN-OR) `cbc` among others (e.g. `gurobi`, `glpk` and a Rust native solver library). As the problem format that is written out and then passed to the solver, it currently only supports the `.lp` file format. Seeing that this format is not so clearly specified, it might make sense to use a more clearly defined format.

I expect that with `CoinUtils` providing the format parsing capabilities of the `COIN-OR` suite, that there is quite some expertise on formats assembled. So my question is, which format would you recommend that:
1. Is well-specified and
2. can be used by most / all solvers?

[I have come across the formats `.nl`, `.mps` and of course `COIN-OR`s `OSIL` format](https://github.com/jcavat/rust-lp-modeler/issues/74). Any recommendations or pointers regarding these formats are welcome!

**Comment by dlaehnemann** [comment]
Not sure how visible this space is. So I'll just ping in @tkralphs and @svigerske as major developers for CoinUtils. And @jjhforrest as a Cbc developer.

My current impression with some further poking around is that:
* both the `.mps` and the `.nl` (AMPL) file format are widely supported, with the `.nl` format more clearly defined (e.g. in [this PDF describing how to write `.nl` files](https://ampl.github.io/nlwrite.pdf))
* `COIN-OR`'s `OSIL` format is by far the most clearly specified format, but is mainly supported by the `COIN-OR` suite

Would that be a reasonable characterization?

**Comment by tkralphs** [comment, marked as answer]
Sorry, I did see this, but didn't get to composing an answer until now. Indeed, `.mps` and `.nl` are widely supported, but `.nl` is generally only used/needed if the instance is non-linear (it was developed original by AMPL to interface solvers with their modeling language, but is an open standard and now widely adopted). 

Despite being an ancient format that should have been replaced long ago, `.mps` is still the option most recommended for MILPs if human readability is not a consideration. The LP format was introduced by CPLEX relatively recently (historically speaking) as a human-readable alternative and has gained popularity and adoption for that reason, but MPS is still a more precisely defined and standardized format that guarantees full precision representations of the problem, among other things. There is a post on Gurobi's support forum that gives a few more details on the advantages of MPS: https://support.gurobi.com/hc/en-us/articles/360013420131

`OSIL` never got much traction and JSON seems to be replacing XML in these kind of use cases. There is a new JSON-based format being developed under the auspices of COIN-OR, which is envisioned as a possible alternative to `.mps` format (https://github.com/coin-modeling-dev).

**Reply by dlaehnemann** [reply]
Ah, perfect. Many thanks for the great overview and on-point insights.

Sounds like `MPS` is the format to turn to currently (with `LP` as the backup for human readability), and the `JSON`-based format like something to look out for in the future.
